packages/react/spec/useFetch.spec.ts

```diff
@@ -93,14 +105,204 @@ describe("useFetch", () => {
+        {
+          "done": false,
+          "value": "hello",
+        }
+      `);
+    });
+    expect(result.current[0].data).toEqual("hello");
+
+    await act(async () => {
+      responseStream.push(" world");
+      expect(await streamReader.read()).toMatchInlineSnapshot(`
+        {
+          "done": false,
+          "value": " world",
+        }
+      `);
+    });
+    expect(result.current[0].data).toEqual("hello world");
+
+    await act(async () => {
+      responseStream.push(null);
+      expect(await streamReader.read()).toMatchInlineSnapshot(`
+        {
+          "done": true,
+          "value": undefined,
+        }
+      `);
+    });
+    expect(result.current[0].data).toEqual("hello world");
+
+    expect(mockUrqlClient[$gadgetConnection].fetch).toBeCalledTimes(1);
+  });
+
+  test("it can fetch a string stream from the backend with a custom encoding", async () => {
+    const { result } = renderHook(() => useFetch("/foo/bar", { stream: "windows-1251", sendImmediately: false }), {
+      wrapper: TestWrapper(relatedProductsApi),
+    });
+
+    expect(result.current[0].data).toBeFalsy();
+    expect(result.current[0].fetching).toBe(false);
+    expect(result.current[0].error).toBeFalsy();
+
+    let sendPromise: Promise<ReadableStream<string>>;
+    act(() => {
+      sendPromise = result.current[1]();
+    });
+
+    expect(result.current[0].data).toBeFalsy();
+    expect(result.current[0].fetching).toBe(true);
+    expect(result.current[0].error).toBeFalsy();
+
+    // eslint-disable-next-line @typescript-eslint/no-empty-function
+    const responseStream = new Readable({ read() {} });
+
+    expect(mockUrqlClient[$gadgetConnection].fetch.requests[0].args).toEqual(["/foo/bar", { sendImmediately: false }]);
+    await mockUrqlClient[$gadgetConnection].fetch.pushResponse(new Response(Readable.toWeb(responseStream) as any));
+
+    let stream: ReadableStream<string>;
+
+    await act(async () => {
+      stream = await sendPromise!;
+    });
+
+    const streamReader = stream!.getReader();
+
+    expect(result.current[0].fetching).toBe(false);
+    expect(result.current[0].error).toBeFalsy();
+
+    expect(result.current[0].data).toEqual("");
+
+    await act(async () => {
+      const bytes = new Uint8Array([207, 240, 232, 226, 229, 242, 44, 32, 236, 232, 240, 33]);
+
+      responseStream.push(bytes);
+      expect(await streamReader.read()).toMatchInlineSnapshot(`
+        {
+          "done": false,
+          "value": "Привет, мир!",
+        }
+      `);
+    });
+
+    expect(result.current[0].data).toEqual("Привет, мир!");
+
+    await act(async () => {
+      responseStream.push(null);
+      expect(await streamReader.read()).toMatchInlineSnapshot(`
+        {
+          "done": true,
+          "value": undefined,
+        }
+      `);
+    });
+    expect(result.current[0].data).toEqual("Привет, мир!");
+
+    expect(mockUrqlClient[$gadgetConnection].fetch).toBeCalledTimes(1);
+  });
```

Comment:
🙏 thanks for putting these together, everyone who goes to touch this is forever in your debt